Customizaiton of Che Cloud Development Environments Blog post

l0rd opened the pull request:
Preview https://pr-l0rd-che-blog.surge.sh/2024/02/05/@mario.loriedo-cde-customization.html

Related to this Dev Spaces issue https://issues.redhat.com/browse/CRW-3211

ibuziuk commented on .devfile.yaml:
```diff
@@ -43,3 +42,6 @@ commands:
         isDefault: false
+events:
+  postStart:
+    - bundle-install
```
no new line

l0rd replied:
thanks, fixed

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### When the default is not enough
+
+The Ruby on Rails Cloud Development Environment that we started in the previous section works for reading the applicaiton source code and do some basic editing but some essential features of the IDE are missing:
```
```suggestion
The Ruby on Rails Cloud Development Environment that we started in the previous section works for reading the application source code and doing some basic editing but some essential features of the IDE are missing:
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+image::/assets/img/cde-customization/ruby-sample-main-branch-ruby-lsp.png[]
+{nbsp} +
+The Ruby VisualStudio Code extension from Shopify fails to activate because Ruby is not installed.
```
```suggestion
The Ruby Visual Studio Code extension from Shopify fails to activate because Ruby is not installed.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+{nbsp} +
+The Ruby VisualStudio Code extension from Shopify fails to activate because Ruby is not installed.
+
+In PART2 we are going to address this problems by using a container image with the required development tools.
```
```suggestion
In PART2 we are going to address those problems by using a container image with the required development tools.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+{nbsp} +
+As a result the image will be built and pushed automatically at every commit. The https://github.com/l0rd/rails-sample/blob/dockerfile/.github/workflows/image-build.yaml#L29[branch `dockerfile` of the ruby-sample repository] pushes the image to `quay.io/mloriedo/rails-blog-cde:latest`.
+
+NOTE: I have named the file as `.devfile.Dockerfile`: even if any name is ok I like to use the `.devfile` prefix as these image are usually used in devfiles (as we will do STEP 3).
```
```suggestion
NOTE: I have named the file as `.devfile.Dockerfile`: even if any name is ok I like to use the `.devfile` prefix as these images are usually used in devfiles (as we will do STEP 3).
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Run Ruby on Rails from the Terminal
+
+And we can now verify that the `rails-sample` application starts successfully from the VisualStudio Code Terminal:
```
```suggestion
And we can now verify that the `rails-sample` application starts successfully from the Visual Studio Code Terminal:
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+image::/assets/img/cde-customization/ruby-lsp-extension-fully-functional.png[]
+{nbsp} +
+Replacing Eclipse Che default image with a custom one allowed to include the right pre-requisites (ruby and rails) to the CDE. With that we are able to run the application and the VisualStudio Code ruby extension. 
```
```suggestion
Replacing Eclipse Che default image with a custom one allowed to include the right pre-requisites (ruby and rails) to the CDE. With that, we are able to run the application and the Visual Studio Code ruby extension. 
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+[TIP]
+====
+In this part the image of the CDE has been replaced using a URL parameter. This can be done using a Devfile too as we will see in next session. An alternative to those approches is to change the image "on the fly" using `kubectl patch` against the DevWorkspace object:
```
```suggestion
In this part the image of the CDE has been replaced using a URL parameter. This can be done using a Devfile too as we will see in the next session. An alternative to those approaches is to change the image "on the fly" using `kubectl patch` against the DevWorkspace object:
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+## PART3: Change CDE properties with a Devfile [[part3]]
+
+In Part 2 we have seen how to customize a CDE with a new container image. That helps to include some development tools that are not in Eclipse Che default environment. But there are other properties of a CDE that a developer may want customize. Properties such as the memory or CPU requirements, containers to run in the CDE Pod (yes, a CDE can have multiple containers!), exposed endpoints, command ...
```
```suggestion
In Part 2 we have seen how to customize a CDE with a new container image. That helps to include some development tools that are not in Eclipse Che default environment. But there are other properties of a CDE that a developer may want to customize. Properties such as the memory or CPU requirements, containers to run in the CDE Pod (yes, a CDE can have multiple containers!), exposed endpoints, commands triggered at startup, at shutdown, etc...These properties can be customized using a Devfile and in this Part 3 we are reviewing how to do that. 
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
````diff
@@ -0,0 +1,434 @@
+      image: quay.io/mloriedo/rails-blog-cde:latest
+```
+
+It's reccomended to add the Devfile in the same git repository of the application developped (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developers provides the git  ...
````
```suggestion
It's recommended to add the Devfile in the same git repository of the application developed (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developer provides the git repository URL to start a CDE.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+TIP: To change the configuration of a running CDE, add a new `.devfile.yaml` (or edit the existing one), and restart the CDE using the command `Eclipse Che: Restart Workspace from Local Devfile` from Visual Studio Code command palette. The command is also quickly accessible by clicking on the bottom-left corner of Visual Studio Code.
+
+TIP: https://github.com/devfile/vscode-walkthrough-extension[The Devfile Visual Studio Code walktrhough extension] helps generating a Devfile using a GUI wizard.
```
```suggestion
TIP: https://github.com/devfile/vscode-walkthrough-extension[The Devfile Visual Studio Code walkthrough extension] helps generating a Devfile using a GUI wizard.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+These steps can be tested iteratively from a running CDE using the "Eclipse Che: Restart Workspace from Local Devfile" in VS Code. And with the YAML extension you get automatic code completion when editing the Devfile.
+
+After last step, the CDE will be fully functional. After the Devfile has been pushed to the git repository any new contributor will be able to run the application, do some changes and debug them in a flash. 
```
```suggestion
After the last step, the CDE will be fully functional. After the Devfile has been pushed to the git repository any new contributor will be able to run the application, do some changes, and debug them in a flash. 
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Add environment variables with Kubernetes `ConfigMaps` and `Secrets`
+
+The Devfile allows to specify environment variables but in some situations you don't want to add them there. Eclipse Che provides a mechanism to automatically add variables to CDEs containers without Devfiles but using Kubernetes ConfigMaps or Secrets.
```
```suggestion
The Devfile allows to specify environment variables but in some situations, you don't want to add them there. Eclipse Che provides a mechanism to automatically add variables to CDEs containers without Devfiles but using Kubernetes ConfigMaps or Secrets.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+The environment variables specified in the `data` field of the `ConfigMap` (in this case `RAILS_DEVELOPMENT_HOSTS=.apps.che-dev.x6e0.p1.openshiftapps.com`) will be added automatically to every CDE created in the same namespace.
+
+An example of environment variables holding sensitive informations are those that specify developers tokens (for example `GITHUB_TOKEN` used by the GitHub CLI).
```
```suggestion
An example of environment variables holding sensitive information are those that specify developers' tokens (for example `GITHUB_TOKEN` used by the GitHub CLI).
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+An example of environment variables holding sensitive informations are those that specify developers tokens (for example `GITHUB_TOKEN` used by the GitHub CLI).
+
+In this case we can create a `Secret` with labels `controller.devfile.io/mount-to-devworkspace: "true"` and `controller.devfile.io/watch-secret: "true"` and with annotation `controller.devfile.io/mount-as: env`:
```
```suggestion
In this case, we can create a `Secret` with labels `controller.devfile.io/mount-to-devworkspace: "true"` and `controller.devfile.io/watch-secret: "true"` and with the annotation `controller.devfile.io/mount-as: env`:
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
````diff
@@ -0,0 +1,434 @@
+EOF
+```
+
+The enviroment variable specified in the `stringData` field of the `Secret` (in this case `GITHUB_TOKEN=<your token>`) will be added automatically to every CDE created in the same namespace. 
````
```suggestion
The environment variable specified in the `stringData` field of the `Secret` (in this case `GITHUB_TOKEN=<your token>`) will be added automatically to every CDE created in the same namespace. 
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+Regular files can be projected into CDEs using `ConfigMaps` and `Secrets` too. In this case the required annotation is `controller.devfile.io/mount-as: file`.
+
+When projecting files, there are two extra annotations read by Eclipse Che: `controller.devfile.io/mount-path: <container-path>` specifes the projected file folder and `controller.devfile.io/mount-access-mode` specifies the projected file access mode.
```
```suggestion
When projecting files, there are two extra annotations read by Eclipse Che: `controller.devfile.io/mount-path: <container-path>` specifies the projected file folder, and `controller.devfile.io/mount-access-mode` specifies the projected file access mode.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+In this last section we want to show a last mechanism to override CDEs default properties. In this case using a https://doc.crds.dev/github.com/devfile/devworkspace-operator[DevWorkspaceOperatorConfig Custom Resource].
+
+`DevWorkspaceOperatorConfig` objects specifies advanced CDE properties such as the `Pod` `schedulerName`, if the `/home/user` folder will be persisted after a restart or not and many more.
```
```suggestion
`DevWorkspaceOperatorConfig` objects specify advanced CDE properties such as the `Pod` `schedulerName`, if the `/home/user` folder will be persisted after a restart or not, and many more.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
````diff
@@ -0,0 +1,434 @@
+...
+```
+
+With that attribute, the CDE started using this devfile, will use the scheduler `my-scheduler` (if it exist) rather then the default Kubernetes scheduler.
````
```suggestion
With that attribute, the CDE started using this devfile will use the scheduler `my-scheduler` (if it exists) rather than the default Kubernetes scheduler.
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+## Conclusion
+
+Most of the time, the Eclipse Che default Cloud Development Environment (CDE) that we have covered in <<part1>>, is good enough to start navigating and do some simple editing of the soure code. 
```
```suggestion
Most of the time, the Eclipse Che default Cloud Development Environment (CDE) that we have covered in <<part1>>, is good enough to start navigating and doing some simple editing of the source code. 
```

ibuziuk commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+To build and run an application in a CDE it may be necessary to create a custom container image and replace the default. This is described in <<part2>>. 
+
+But it's in <<part3>> that we shared how to make to most out of Eclipse Che development environments. The Devfile allow to specify development environments as code and share it with the rest of the team. 
```
```suggestion
But it's in <<part3>> that we shared how to make to most out of Eclipse Che development environments. The Devfile allows us to specify development environments as code and share it with the rest of the team. 
```

ibuziuk reviewed (approved):
top content and great read!

l0rd commented:
I have addressed all comments. I have also added a devfile command to publish the blog content to surge.sh (waiting for the GitHub workflow to be addressed).

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+## PART 1: Eclipse Che default CDE [[part1]]
+
+When no Devfile is found in a git repository and the developer doesn't provide an URL, then Eclipse Che uses a default development environment.
```
```suggestion
When no Devfile is found in a git repository and the developer doesn't provide a URL, Eclipse Che uses a default development environment.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+<<part4>>
+
+
+NOTE: To help illustrate this sample we will use a https://github.com/l0rd/rails-sample[Ruby on Rails source code example]. Ruby is a good example because the default development container doesn't include Ruby development tools at all. For other languages such as Java, Go, NodeJS, Python and .NET, Eclispe Che default development image includes a few tools, and the customizations in this blog post ...
```
```suggestion
NOTE: To help illustrate this sample we will use a https://github.com/l0rd/rails-sample[Ruby on Rails source code example]. Ruby is a good example because the default development container doesn't include Ruby development tools at all. For other languages such as Java, Go, NodeJS, Python and .NET, the Eclipse Che default development image includes a few tools, and the customizations in this blog post may not be necessary.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+slug: /@mario.loriedo/cde-customization
+---
+
+Eclipse Che provides Cloud Development Environments (CDEs) with a click. The default container of those CDEs, what we call the https://github.com/devfile/developer-images/[universal developer image], has a lot of pre-installed development tools but may not have the right version or the right tool for you. If that's the case, no worries: one Eclipse Che CDEs main assets is configurability! 
```
```suggestion
Eclipse Che provides Cloud Development Environments (CDEs) with a click. The default container of those CDEs, what we call the https://github.com/devfile/developer-images/[universal developer image], has a lot of pre-installed development tools but may not have the right version or the right tool for you. If that's the case, no worries: one of Eclipse Che CDEs' main assets is configurability! 
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+Eclipse Che provides Cloud Development Environments (CDEs) with a click. The default container of those CDEs, what we call the https://github.com/devfile/developer-images/[universal developer image], has a lot of pre-installed development tools but may not have the right version or the right tool for you. If that's the case, no worries: one Eclipse Che CDEs main assets is configurability! 
+
+In this blog posts we will review Eclipse Che CDEs customization in 4 parts:
```
```suggestion
In these blog posts we will review Eclipse Che CDEs' customization in 4 parts:
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+> Devfile could not be found in https://github.com/l0rd/rails-sample. Applying the default configuration
+
+NOTE: You can click https://workspaces.openshift.com/#https://github.com/l0rd/rails-sample[here] to open https://github.com/l0rd/rails-sample[] in Red Hat Developer Sandbox and try it out yourself.
```
```suggestion
NOTE: You can click https://workspaces.openshift.com/#https://github.com/l0rd/rails-sample[here] to open https://github.com/l0rd/rails-sample[] in the Red Hat Developer Sandbox and try it out yourself.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Default CDE Pod details
+
+Looking more closely we can see that, once the CDE is started, a new Pod is created.
```
```suggestion
Looking more closely we can see that once the CDE is started, a new Pod is created.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+For simplicity the base image is Eclipse Che Universal Developer Image (`quay.io/devfile/universal-developer-image:latest`) and `ruby` and `rails` are installed on top of it.
+
+NOTE: It's not necessary to use Eclipe Che default image as the base image. I am maintaining https://github.com/devfile/developer-images/[a repository with a list of Dockerfiles for Che CDEs] using popular base images.
```
```suggestion
NOTE: It's not necessary to use the Eclipse Che default image as the base image. I am maintaining https://github.com/devfile/developer-images/[a repository with a list of Dockerfiles for Che CDEs] using popular base images.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+{nbsp} +
+As a result the image will be built and pushed automatically at every commit. The https://github.com/l0rd/rails-sample/blob/dockerfile/.github/workflows/image-build.yaml#L29[branch `dockerfile` of the ruby-sample repository] pushes the image to `quay.io/mloriedo/rails-blog-cde:latest`.
+
+NOTE: I have named the file as `.devfile.Dockerfile`: even if any name is ok I like to use the `.devfile` prefix as these images are usually used in devfiles (as we will do STEP 3).
```
```suggestion
NOTE: I have named the file as `.devfile.Dockerfile`: even though any name is ok I like to use the `.devfile` prefix as these images are usually used in devfiles (as we will do STEP 3).
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+image::/assets/img/cde-customization/image-url-parameter.png[]
+{nbsp} +
+
+Or this URL can be used to start it on Red Hat Developer Sandbox: 
```
```suggestion
Or this URL can be used to start it on the Red Hat Developer Sandbox: 
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Verify the new CDE Pod image
+
+The CDE Pod uses a `rails-blog-cde` container rather than Eclipse Che default universal developer image. To verify it open a Terminal from VisualStudio Code and run kubectl:
```
```suggestion
The CDE Pod now uses a `rails-blog-cde` container rather than the Eclipse Che default universal developer image. To verify it open a Terminal from Visual Studio Code and run `kubectl`:
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Install `ruby-lsp` Visual Studio Code extension
+
+Finally, the Ruby LSP extension, that failed to activate in STEP1, is now fully working:
```
```suggestion
Finally, the Ruby LSP extension, which failed to activate in STEP1, is now fully working:
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+image::/assets/img/cde-customization/ruby-lsp-extension-fully-functional.png[]
+{nbsp} +
+Replacing Eclipse Che default image with a custom one allowed to include the right pre-requisites (ruby and rails) to the CDE. With that, we are able to run the application and the Visual Studio Code ruby extension. 
```
```suggestion
Replacing the Eclipse Che default image with a custom one allowed us to include the right pre-requisites (ruby and rails) to the CDE. With that, we are able to run the application and the Visual Studio Code ruby extension. 
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+{nbsp} +
+Replacing Eclipse Che default image with a custom one allowed to include the right pre-requisites (ruby and rails) to the CDE. With that, we are able to run the application and the Visual Studio Code ruby extension. 
+
+There are still a few properties of the CDE that can be improved (like increasing the CPU and memory, specifying the command to run the application, pre-fetch the dependencies etc...). We will cover those in the next session where we will use the Devfile for CDE more advanced configurations.
```
```suggestion
There are still a few properties of the CDE that can be improved (like increasing the CPU and memory, specifying the command to run the application, pre-fetch the dependencies etc...). We will cover those in the next session where we will use the Devfile to perform more advanced configurations for the CDE.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+## PART3: Change CDE properties with a Devfile [[part3]]
+
+In Part 2 we have seen how to customize a CDE with a new container image. That helps to include some development tools that are not in Eclipse Che default environment. But there are other properties of a CDE that a developer may want to customize. Properties such as the memory or CPU requirements, containers to run in the CDE Pod (yes, a CDE can have multiple containers!), exposed endpoints, comm ...
```
```suggestion
In Part 2 we have seen how to customize a CDE with a new container image. That helps include some development tools that are not in Eclipse Che default environment. But there are other properties of a CDE that a developer may want to customize. Properties such as the memory or CPU requirements, containers to run in the CDE Pod (yes, a CDE can have multiple containers!), exposed endpoints, commands triggered at startup, at shutdown, etc...These properties can be customized using a Devfile and in this Part 3 we are reviewing how to do that. 
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Define a Devfile for an Eclipse Che CDE
+
+A Devfile it's an https://www.cncf.io/projects/devfile/[open standard] to define Cloud Development Environments. It's a YAML file used to configure Eclipse Che CDEs.
```
```suggestion
A Devfile is an https://www.cncf.io/projects/devfile/[open standard] to define Cloud Development Environments. It's a YAML file used to configure Eclipse Che CDEs.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+A Devfile it's an https://www.cncf.io/projects/devfile/[open standard] to define Cloud Development Environments. It's a YAML file used to configure Eclipse Che CDEs.
+
+This is a simple Devfile that specifies the container image built in STEP 2 looks like:
```
```suggestion
This is a simple Devfile that specifies the container image built in STEP 2:
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
````diff
@@ -0,0 +1,434 @@
+      image: quay.io/mloriedo/rails-blog-cde:latest
+```
+
+It's recommended to add the Devfile in the same git repository of the application developed (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developer provides the git re ...
````
```suggestion
It's recommended to add the Devfile in the same git repository of the application being developed (as seen in the https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developer provides the git repository URL to start a CDE.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+It's recommended to add the Devfile in the same git repository of the application developed (as https://github.com/l0rd/rails-sample/tree/devfile[the `devfile` branch of the `rails-sample` git repository]) using the name `.devfile.yaml`. When the Devfile is in the git repository it will evolve along with the application and it will be applied automatically whenever a developer provides the git re ...
+
+When it's not possible to add a `.devfile.yaml` in the git repository, there is still the option to publish it (in a gist, pastebin or another git repository) and https://eclipse.dev/che/docs/stable/end-user-guide/starting-a-workspace-from-a-raw-devfile-url/[start the CDE providing the raw URL to it].
```
```suggestion
When it's not possible to add a `.devfile.yaml` in the git repository, there is still the option to publish it (in a gist, pastebin or another git repository) and https://eclipse.dev/che/docs/stable/end-user-guide/starting-a-workspace-from-a-raw-devfile-url/[start the CDE by providing the raw URL to it].
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+When it's not possible to add a `.devfile.yaml` in the git repository, there is still the option to publish it (in a gist, pastebin or another git repository) and https://eclipse.dev/che/docs/stable/end-user-guide/starting-a-workspace-from-a-raw-devfile-url/[start the CDE providing the raw URL to it].
+
+TIP: To change the configuration of a running CDE, add a new `.devfile.yaml` (or edit the existing one), and restart the CDE using the command `Eclipse Che: Restart Workspace from Local Devfile` from Visual Studio Code command palette. The command is also quickly accessible by clicking on the bottom-left corner of Visual Studio Code.
```
```suggestion
TIP: To change the configuration of a running CDE, add a new `.devfile.yaml` (or edit the existing one), and restart the CDE using the command `Eclipse Che: Restart Workspace from Local Devfile` from the Visual Studio Code command palette. This command is also quickly accessible by clicking on the bottom-left corner of Visual Studio Code.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+++++
+{nbsp} +
+
+These steps can be tested iteratively from a running CDE using the "Eclipse Che: Restart Workspace from Local Devfile" in VS Code. And with the YAML extension you get automatic code completion when editing the Devfile.
```
```suggestion
These steps can be tested iteratively from a running CDE using the "Eclipse Che: Restart Workspace from Local Devfile" command in VS Code. And with the YAML extension you get automatic code completion when editing the Devfile.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+After the last step, the CDE will be fully functional. After the Devfile has been pushed to the git repository any new contributor will be able to run the application, do some changes, and debug them in a flash. 
+
+For example you can use https://workspaces.openshift.com/#https://github.com/l0rd/rails-sample/tree/devfile[this link] to test the rails-sample using the Devfile in step 7. After the workspace has started, use task `server-start` in VS Code (Terminal => Run Tasks => Devfile) to start the application and install the VS extension `ruby-lsp` to start making changes with full language support.
```
```suggestion
For example you can use https://workspaces.openshift.com/#https://github.com/l0rd/rails-sample/tree/devfile[this link] to test the rails-sample using the Devfile in step 7. After the workspace has started, use task `server-start` in VS Code (Terminal => Run Tasks => Devfile) to start the application and install the VS Code extension `ruby-lsp` to start making changes with full language support.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+## PART4: Secrets and other CDE configurations that cannot be specified in Devfile [[part4]]
+
+There are some properties of your CDE that you don't want to specify in a Devfile. Either because they contain sensible information (like a password or an SSH private key), because you want to customize your CDE without affecting the rest of the team or because that's an IDE configuration. In any case, the Devfile that is a shared file in the git repository, cannot be used. In this section we wil ...
```
```suggestion
There are some properties of your CDE that you don't want to specify in a Devfile. Either because they contain sensitive information (like a password or an SSH private key), because you want to customize your CDE without affecting the rest of the team or because it's an IDE configuration. In any case, the Devfile which is a shared file in the git repository, cannot be used. In this section we will go through a few techniques to add user specific configurations.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+The environment variable specified in the `stringData` field of the `Secret` (in this case `GITHUB_TOKEN=<your token>`) will be added automatically to every CDE created in the same namespace. 
+
+TIP: These ConfigMap and Secret should be created in the developer namespace. The variables specified in `data` will be added to the CDEs of owner of the namespace only. Dev Spaces uses namespaces to isolate developers CDEs: it's critical to allow read access to the objects in the namespaces only to the developer owning it.
```
```suggestion
TIP: These ConfigMap and Secret should be created in the developer namespace. The variables specified in `data` will be added to the CDEs of the owner of the namespace only. Eclipse Che uses namespaces to isolate developers CDEs: it's critical to allow read access to the objects in the namespaces only to the developer owning it.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+Regular files can be projected into CDEs using `ConfigMaps` and `Secrets` too. In this case the required annotation is `controller.devfile.io/mount-as: file`.
+
+When projecting files, there are two extra annotations read by Eclipse Che: `controller.devfile.io/mount-path: <container-path>` specifies the projected file folder, and `controller.devfile.io/mount-access-mode` specifies the projected file access mode.
```
```suggestion
When projecting files, there are two extra annotations read by Eclipse Che: `controller.devfile.io/mount-path: <container-path>` specifies the projected file folder, and `controller.devfile.io/mount-access-mode` specifies the projected file's access mode.
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+### Override defaults configurations using a DevWorkspace Operator Configuration
+
+In this last section we want to show a last mechanism to override CDEs default properties. In this case using a https://doc.crds.dev/github.com/devfile/devworkspace-operator[DevWorkspaceOperatorConfig Custom Resource].
```
```suggestion
In this last section we want to show a last mechanism to override CDEs' default properties. In this case, by using a https://doc.crds.dev/github.com/devfile/devworkspace-operator[DevWorkspaceOperatorConfig Custom Resource].
```

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+`DevWorkspaceOperatorConfig` objects specify advanced CDE properties such as the `Pod` `schedulerName`, whether the `/home/user` folder will persist after a restart or not, and many more.
+
+Applying a particular `DevWorkspaceOperatorConfig` to a CDEs takes 2 steps:
+1. Create the `DevWorkspaceOperatorConfig` custom resource in the Kubernetes cluster
```
```suggestion

1. Create the `DevWorkspaceOperatorConfig` custom resource in the Kubernetes cluster
```

dkwon17 replied:
The numbered list is not being displayed correctly, I think the newline should fix it:

<img width="786" alt="image" src="https://github.com/eclipse-che/blog/assets/83611742/36f4ff1a-0204-4e4c-8e74-b0507018c225">

dkwon17 commented on _posts/2024-02-05-cde-customization.adoc:
```diff
@@ -0,0 +1,434 @@
+
+Applying a particular `DevWorkspaceOperatorConfig` to a CDEs takes 2 steps:
+1. Create the `DevWorkspaceOperatorConfig` custom resource in the Kubernetes cluster
+2. Edit the Devfile to specify to apply that specific configuration
```
```suggestion
2. Edit the Devfile to apply that specific configuration
```

l0rd commented:
Thank you @dkwon17 for the review. I have accepted your suggestions (all).

dkwon17 reviewed (approved):
No problem, it's a great blog post!
